src/tracely/clean_trace.py

```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+            trace_2 (np.array): The second trace, represented as a NumPy array, where each row is [latitude, longitude, timestamp].
+            distance_threshold (float): The maximum allowable spatial distance (in meters) for two pings to be considered overlapping.
+            time_threshold (int): The maximum allowable time difference (in the same units as the timestamps, eg. unix epoch timestamp) for two pings to be considered overlapping.
+
+        Returns:
+            dict: 
+                A dictionary containing:
+                    similarity_percentage (float): The percentage of pings in `trace_1` that have overlapping pings in `trace_2`.
+                    overlapping_pings_indices (list): A list of pairs of indices, where each pair represents the indices of overlapping pings in `trace_1` and `trace_2`.
+
+        Notes:
+            - The distance between pings is calculated using the Haversine formula.
+
+        """
+
+        # Create KDTree for time filtering on trace_2
+        time_pings_2 = trace_2[:, 2].reshape(-1, 1)  # Extract timestamps as 1D pings
+        time_tree_2 = KDTree(time_pings_2)
+
+        # Initialize variables
+        matched_pings_counts = 0
+        overlapping_pings_pairs = []
+
+        # Iterate through pings in trace_1
+        for i, (lat1, lon1, ts1) in enumerate(trace_1):
+            # Query pings in trace_2 within the time threshold
+            candidates = time_tree_2.query_ball_point(ts1, r=time_threshold)
+
+            if not candidates:
+                continue
+
+            # Extract spatial pings of candidates
+            candidate_pings = trace_2[candidates, :2]
+
+            # Calculate the difference array
+            candidate_pings_np = np.array(candidate_pings)
+            difference_array = candidate_pings_np - np.array([lat1, lon1])
+            difference_sums = np.sum(np.abs(difference_array), axis=1)
+
+            # Find the closest candidate by minimum sum of differences
+            closest_index = np.argmin(difference_sums)
+            closest_point = candidate_pings[closest_index]
+
+            # Check Haversine distance
+            lat2, lon2 = closest_point
+            haversine_distance = get_haversine_distance(lat1, lon1, lat2, lon2)
+
+            if haversine_distance <= distance_threshold:
+                matched_pings_counts += 1
+                overlapping_pings_pairs.append([i, candidates[closest_index]])
+
+        # Calculate similarity percentage
+        similarity_percentage = (matched_pings_counts / len(trace_1)) * 100
+        similarity_percentage = round(similarity_percentage, 5)
+
+        # Return results
+        return {
+            "similarity_percentage": similarity_percentage,
+            "overlapping_pings_indices": overlapping_pings_pairs
+        }
+
+    @staticmethod
+    def calculate_trace_similarity(trace_1, trace_2, distance_threshold, time_threshold, plot_map=False):
```

Comment:
This function does not belong to the clean trace class. Let's create a package called "trace_functions" or something and within it we can have "trace_similarity.py" module. This leaves room for adding multiple functionalities here itself.